# Chapter 5 - Exercise 2: Giao dịch chứng khoán

In [4]:
from google.colab import drive

drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


##### Cho 3 file .csv sau:
* **stocks1.csv :** *date*, *symbol*, *open*, *high*, *low*, *close*, *volume* : chứa thông tin giao dịch chứng khoán các công ty khác nhau
* **stocks2.csv :** *date*, *symbol*, *open*, *high*, *low*, *close*, *volume* : chứa thông tin giao dịch chứng khoán các công ty khác nhau
* **companies.csv :** *name*, *employees*, *headquarters_city*, *headquarters_state* : chứa thông tin về trụ sở và số lượng nhân viên cho một công ty cụ thể

#### Thực hiện các yêu cầu sau, và đối chiếu với kết quả cho trước:

In [5]:
import pandas as pd
import numpy as np

In [6]:
# Câu 1a: Đọc file stocks1.csv => đưa dữ liệu vào stocks1
# Câu 1a: Đọc file stocks1.csv => đưa dữ liệu vào stocks1
stocks1 = pd.read_csv('/content/gdrive/My Drive/stocks1.csv')
# Hiển thị 5 dòng dữ liệu đầu của stocks1
stocks1.head()
# Hiển thị 5 dòng dữ liệu đầu của stocks1


date symbol     open     high      low    close   volume
0  01-03-19   AMZN  1655.13  1674.26  1651.00  1671.73  4974877
1  04-03-19   AMZN  1685.00  1709.43  1674.36  1696.17  6167358
2  05-03-19   AMZN  1702.95  1707.80  1689.01  1692.43  3681522
3  06-03-19   AMZN  1695.97      NaN      NaN  1668.95  3996001
4  07-03-19   AMZN  1667.37  1669.75  1620.51  1625.95  4957017

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>AMZN</td>
      <td>1655.13</td>
      <td>1674.26</td>
      <td>1651.00</td>
      <td>1671.73</td>
      <td>4974877</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>AMZN</td>
      <td>1685.00</td>
      <td>1709.43</td>
      <td>1674.36</td>
      <td>1696.17</td>
      <td>6167358</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>AMZN</td>
      <td>1702.95</td>
      <td>1707.80</td>
      <td>1689.01</td>
      <td>1692.43</td>
      <td>3681522</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>AMZN</td>
      <td>1695.97</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>1668.95</td>
      <td>3996001</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>AMZN</td>
      <td>1667.37</td>
      <td>1669.75</td>
      <td>1620.51</td>
      <td>1625.95</td>
      <td>4957017</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Hiển thị 5 dòng dữ liệu cuối của stocks1


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>01-03-19</td>
      <td>GOOG</td>
      <td>1124.90</td>
      <td>1142.97</td>
      <td>1124.75</td>
      <td>1140.99</td>
      <td>1450316</td>
    </tr>
    <tr>
      <th>11</th>
      <td>04-03-19</td>
      <td>GOOG</td>
      <td>1146.99</td>
      <td>1158.28</td>
      <td>1130.69</td>
      <td>1147.80</td>
      <td>1446047</td>
    </tr>
    <tr>
      <th>12</th>
      <td>05-03-19</td>
      <td>GOOG</td>
      <td>1150.06</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>1162.03</td>
      <td>1443174</td>
    </tr>
    <tr>
      <th>13</th>
      <td>06-03-19</td>
      <td>GOOG</td>
      <td>1162.49</td>
      <td>1167.57</td>
      <td>1155.49</td>
      <td>1157.86</td>
      <td>1099289</td>
    </tr>
    <tr>
      <th>14</th>
      <td>07-03-19</td>
      <td>GOOG</td>
      <td>1155.72</td>
      <td>1156.76</td>
      <td>1134.91</td>
      <td>1143.30</td>
      <td>1166559</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Cho biết kiểu dữ liệu (dtype) của các cột của stocks1


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>date       object
symbol     object
open      float64
high      float64
low       float64
close     float64
volume      int64
dtype: object</pre>
  
</details>

In [ ]:
# Xem thông tin (info) của stocks1


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 15 entries, 0 to 14
Data columns (total 7 columns):
date      15 non-null object
symbol    15 non-null object
open      15 non-null float64
high      13 non-null float64
low       13 non-null float64
close     15 non-null float64
volume    15 non-null int64
dtypes: float64(4), int64(1), object(2)
memory usage: 968.0+ bytes
</pre>
  
</details>

In [ ]:
# Câu 1b: Đọc file stocks2.csv => đưa dữ liệu vào stocks2

# Hiển thị 5 dòng dữ liệu đầu của stocks2


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>FB</td>
      <td>162.60</td>
      <td>163.132</td>
      <td>161.69</td>
      <td>162.28</td>
      <td>11097770</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>FB</td>
      <td>163.90</td>
      <td>167.500</td>
      <td>163.83</td>
      <td>167.37</td>
      <td>18894689</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>FB</td>
      <td>167.37</td>
      <td>171.880</td>
      <td>166.55</td>
      <td>171.26</td>
      <td>28187890</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>FB</td>
      <td>172.90</td>
      <td>173.570</td>
      <td>171.27</td>
      <td>172.51</td>
      <td>21531723</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>FB</td>
      <td>171.50</td>
      <td>171.740</td>
      <td>167.61</td>
      <td>169.13</td>
      <td>18306504</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Hiển thị 5 dòng dữ liệu cuối của stocks2


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>5</th>
      <td>01-03-19</td>
      <td>TSLA</td>
      <td>306.94</td>
      <td>307.1300</td>
      <td>291.90</td>
      <td>294.79</td>
      <td>22911375</td>
    </tr>
    <tr>
      <th>6</th>
      <td>04-03-19</td>
      <td>TSLA</td>
      <td>298.12</td>
      <td>299.0000</td>
      <td>282.78</td>
      <td>285.36</td>
      <td>17096818</td>
    </tr>
    <tr>
      <th>7</th>
      <td>05-03-19</td>
      <td>TSLA</td>
      <td>282.00</td>
      <td>284.0000</td>
      <td>270.10</td>
      <td>276.54</td>
      <td>18764740</td>
    </tr>
    <tr>
      <th>8</th>
      <td>06-03-19</td>
      <td>TSLA</td>
      <td>276.48</td>
      <td>281.5058</td>
      <td>274.39</td>
      <td>276.24</td>
      <td>10335485</td>
    </tr>
    <tr>
      <th>9</th>
      <td>07-03-19</td>
      <td>TSLA</td>
      <td>278.84</td>
      <td>284.7000</td>
      <td>274.25</td>
      <td>276.59</td>
      <td>9442483</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Cho biết kiểu dữ liệu (dtype) của các cột của stocks2


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>date       object
symbol     object
open      float64
high      float64
low       float64
close     float64
volume      int64
dtype: object</pre>
  
</details>

In [ ]:
# Xem thông tin (info) của stocks2


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 10 entries, 0 to 9
Data columns (total 7 columns):
date      10 non-null object
symbol    10 non-null object
open      10 non-null float64
high      10 non-null float64
low       10 non-null float64
close     10 non-null float64
volume    10 non-null int64
dtypes: float64(4), int64(1), object(2)
memory usage: 688.0+ bytes
</pre>
  
</details>

In [ ]:
# Câu 1c: Đọc file companies.csv => đưa dữ liệu vào companies

# Xem dữ liệu của companies


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>employees</th>
      <th>headquarters_city</th>
      <th>headquarters_state</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GOOG</td>
      <td>98771</td>
      <td>Mountain View</td>
      <td>CA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AAPL</td>
      <td>132000</td>
      <td>Cupertino</td>
      <td>CA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>FB</td>
      <td>48268</td>
      <td>Menlo Park</td>
      <td>CA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TSLA</td>
      <td>48016</td>
      <td>Palo Alto</td>
      <td>CA</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Cho biết kiểu dữ liệu (dtype) của các cột của companies


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>name                  object
employees              int64
headquarters_city     object
headquarters_state    object
dtype: object</pre>
  
</details>

In [ ]:
# Xem thông tin (info) của companies


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
name                  5 non-null object
employees             5 non-null int64
headquarters_city     5 non-null object
headquarters_state    5 non-null object
dtypes: int64(1), object(3)
memory usage: 288.0+ bytes
</pre>
  
</details>

In [ ]:
# Câu 2: Cho biết trong stocks1 có dữ liệu Null hay không?  


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>date      False
symbol    False
open      False
high       True
low        True
close     False
volume    False
dtype: bool</pre>
  
</details>

In [26]:
# Nếu có, hãy thay thế với quy tắc sau:
# Nếu Null cột 'high' thì thay bằng giá trị max trên cột 'high' của mã chứng khoán đó
# Nếu Null cột 'low' thì thay bằng giá trị min trên cột 'low' của mã chứng khoán đó

stocks1['high'] = stocks1['high'].fillna(stocks1.groupby('symbol')['high'].transform(max))
    
stocks1['low'] = stocks1['low'].fillna(stocks1.groupby('symbol')['low'].transform(min))
    
stocks1


date symbol     open     high      low    close    volume
0   01-03-19   AMZN  1655.13  1674.26  1651.00  1671.73   4974877
1   04-03-19   AMZN  1685.00  1709.43  1674.36  1696.17   6167358
2   05-03-19   AMZN  1702.95  1707.80  1689.01  1692.43   3681522
3   06-03-19   AMZN  1695.97  1709.43  1620.51  1668.95   3996001
4   07-03-19   AMZN  1667.37  1669.75  1620.51  1625.95   4957017
5   01-03-19   AAPL   174.28   175.15   172.89   174.97  25886167
6   04-03-19   AAPL   175.69   177.75   173.97   175.85  27436203
7   05-03-19   AAPL   175.94   176.00   174.54   175.53  19737419
8   06-03-19   AAPL   174.67   175.49   173.94   174.52  20810384
9   07-03-19   AAPL   173.87   174.44   172.02   172.50  24796374
10  01-03-19   GOOG  1124.90  1142.97  1124.75  1140.99   1450316
11  04-03-19   GOOG  1146.99  1158.28  1130.69  1147.80   1446047
12  05-03-19   GOOG  1150.06  1167.57  1124.75  1162.03   1443174
13  06-03-19   GOOG  1162.49  1167.57  1155.49  1157.86   1099289
14  07-03-19   GOOG  1155.72  1156.76  1134.91  1143.30   1166559

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>AMZN</td>
      <td>1655.13</td>
      <td>1674.26</td>
      <td>1651.00</td>
      <td>1671.73</td>
      <td>4974877</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>AMZN</td>
      <td>1685.00</td>
      <td>1709.43</td>
      <td>1674.36</td>
      <td>1696.17</td>
      <td>6167358</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>AMZN</td>
      <td>1702.95</td>
      <td>1707.80</td>
      <td>1689.01</td>
      <td>1692.43</td>
      <td>3681522</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>AMZN</td>
      <td>1695.97</td>
      <td>1709.43</td>
      <td>1620.51</td>
      <td>1668.95</td>
      <td>3996001</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>AMZN</td>
      <td>1667.37</td>
      <td>1669.75</td>
      <td>1620.51</td>
      <td>1625.95</td>
      <td>4957017</td>
    </tr>
    <tr>
      <th>5</th>
      <td>01-03-19</td>
      <td>AAPL</td>
      <td>174.28</td>
      <td>175.15</td>
      <td>172.89</td>
      <td>174.97</td>
      <td>25886167</td>
    </tr>
    <tr>
      <th>6</th>
      <td>04-03-19</td>
      <td>AAPL</td>
      <td>175.69</td>
      <td>177.75</td>
      <td>173.97</td>
      <td>175.85</td>
      <td>27436203</td>
    </tr>
    <tr>
      <th>7</th>
      <td>05-03-19</td>
      <td>AAPL</td>
      <td>175.94</td>
      <td>176.00</td>
      <td>174.54</td>
      <td>175.53</td>
      <td>19737419</td>
    </tr>
    <tr>
      <th>8</th>
      <td>06-03-19</td>
      <td>AAPL</td>
      <td>174.67</td>
      <td>175.49</td>
      <td>173.94</td>
      <td>174.52</td>
      <td>20810384</td>
    </tr>
    <tr>
      <th>9</th>
      <td>07-03-19</td>
      <td>AAPL</td>
      <td>173.87</td>
      <td>174.44</td>
      <td>172.02</td>
      <td>172.50</td>
      <td>24796374</td>
    </tr>
    <tr>
      <th>10</th>
      <td>01-03-19</td>
      <td>GOOG</td>
      <td>1124.90</td>
      <td>1142.97</td>
      <td>1124.75</td>
      <td>1140.99</td>
      <td>1450316</td>
    </tr>
    <tr>
      <th>11</th>
      <td>04-03-19</td>
      <td>GOOG</td>
      <td>1146.99</td>
      <td>1158.28</td>
      <td>1130.69</td>
      <td>1147.80</td>
      <td>1446047</td>
    </tr>
    <tr>
      <th>12</th>
      <td>05-03-19</td>
      <td>GOOG</td>
      <td>1150.06</td>
      <td>1167.57</td>
      <td>1124.75</td>
      <td>1162.03</td>
      <td>1443174</td>
    </tr>
    <tr>
      <th>13</th>
      <td>06-03-19</td>
      <td>GOOG</td>
      <td>1162.49</td>
      <td>1167.57</td>
      <td>1155.49</td>
      <td>1157.86</td>
      <td>1099289</td>
    </tr>
    <tr>
      <th>14</th>
      <td>07-03-19</td>
      <td>GOOG</td>
      <td>1155.72</td>
      <td>1156.76</td>
      <td>1134.91</td>
      <td>1143.30</td>
      <td>1166559</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 3: Tạo dataframe stocks bằng cách gộp stocks1 và stocks2 theo dòng

# Xem 15 dòng dữ liệu cuối của stocks


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>01-03-19</td>
      <td>GOOG</td>
      <td>1124.90</td>
      <td>1142.9700</td>
      <td>1124.75</td>
      <td>1140.99</td>
      <td>1450316</td>
    </tr>
    <tr>
      <th>11</th>
      <td>04-03-19</td>
      <td>GOOG</td>
      <td>1146.99</td>
      <td>1158.2800</td>
      <td>1130.69</td>
      <td>1147.80</td>
      <td>1446047</td>
    </tr>
    <tr>
      <th>12</th>
      <td>05-03-19</td>
      <td>GOOG</td>
      <td>1150.06</td>
      <td>1167.5700</td>
      <td>1124.75</td>
      <td>1162.03</td>
      <td>1443174</td>
    </tr>
    <tr>
      <th>13</th>
      <td>06-03-19</td>
      <td>GOOG</td>
      <td>1162.49</td>
      <td>1167.5700</td>
      <td>1155.49</td>
      <td>1157.86</td>
      <td>1099289</td>
    </tr>
    <tr>
      <th>14</th>
      <td>07-03-19</td>
      <td>GOOG</td>
      <td>1155.72</td>
      <td>1156.7600</td>
      <td>1134.91</td>
      <td>1143.30</td>
      <td>1166559</td>
    </tr>
    <tr>
      <th>15</th>
      <td>01-03-19</td>
      <td>FB</td>
      <td>162.60</td>
      <td>163.1320</td>
      <td>161.69</td>
      <td>162.28</td>
      <td>11097770</td>
    </tr>
    <tr>
      <th>16</th>
      <td>04-03-19</td>
      <td>FB</td>
      <td>163.90</td>
      <td>167.5000</td>
      <td>163.83</td>
      <td>167.37</td>
      <td>18894689</td>
    </tr>
    <tr>
      <th>17</th>
      <td>05-03-19</td>
      <td>FB</td>
      <td>167.37</td>
      <td>171.8800</td>
      <td>166.55</td>
      <td>171.26</td>
      <td>28187890</td>
    </tr>
    <tr>
      <th>18</th>
      <td>06-03-19</td>
      <td>FB</td>
      <td>172.90</td>
      <td>173.5700</td>
      <td>171.27</td>
      <td>172.51</td>
      <td>21531723</td>
    </tr>
    <tr>
      <th>19</th>
      <td>07-03-19</td>
      <td>FB</td>
      <td>171.50</td>
      <td>171.7400</td>
      <td>167.61</td>
      <td>169.13</td>
      <td>18306504</td>
    </tr>
    <tr>
      <th>20</th>
      <td>01-03-19</td>
      <td>TSLA</td>
      <td>306.94</td>
      <td>307.1300</td>
      <td>291.90</td>
      <td>294.79</td>
      <td>22911375</td>
    </tr>
    <tr>
      <th>21</th>
      <td>04-03-19</td>
      <td>TSLA</td>
      <td>298.12</td>
      <td>299.0000</td>
      <td>282.78</td>
      <td>285.36</td>
      <td>17096818</td>
    </tr>
    <tr>
      <th>22</th>
      <td>05-03-19</td>
      <td>TSLA</td>
      <td>282.00</td>
      <td>284.0000</td>
      <td>270.10</td>
      <td>276.54</td>
      <td>18764740</td>
    </tr>
    <tr>
      <th>23</th>
      <td>06-03-19</td>
      <td>TSLA</td>
      <td>276.48</td>
      <td>281.5058</td>
      <td>274.39</td>
      <td>276.24</td>
      <td>10335485</td>
    </tr>
    <tr>
      <th>24</th>
      <td>07-03-19</td>
      <td>TSLA</td>
      <td>278.84</td>
      <td>284.7000</td>
      <td>274.25</td>
      <td>276.59</td>
      <td>9442483</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 4: Tạo dataframe stocks_companies bằng cách gộp stocks và companies 

# Xem 5 dòng dữ liệu đầu của stocks_companies


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
      <th>name</th>
      <th>employees</th>
      <th>headquarters_city</th>
      <th>headquarters_state</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>AMZN</td>
      <td>1655.13</td>
      <td>1674.26</td>
      <td>1651.00</td>
      <td>1671.73</td>
      <td>4974877</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>AMZN</td>
      <td>1685.00</td>
      <td>1709.43</td>
      <td>1674.36</td>
      <td>1696.17</td>
      <td>6167358</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>AMZN</td>
      <td>1702.95</td>
      <td>1707.80</td>
      <td>1689.01</td>
      <td>1692.43</td>
      <td>3681522</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>AMZN</td>
      <td>1695.97</td>
      <td>1709.43</td>
      <td>1620.51</td>
      <td>1668.95</td>
      <td>3996001</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>AMZN</td>
      <td>1667.37</td>
      <td>1669.75</td>
      <td>1620.51</td>
      <td>1625.95</td>
      <td>4957017</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 5: Cho biết giá (open, high, low, close) trung bình  và volume trung bình của mỗi công ty


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
    </tr>
    <tr>
      <th>symbol</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAPL</th>
      <td>174.890</td>
      <td>175.76600</td>
      <td>173.472</td>
      <td>174.674</td>
      <td>23733309.4</td>
    </tr>
    <tr>
      <th>AMZN</th>
      <td>1681.284</td>
      <td>1694.13400</td>
      <td>1651.078</td>
      <td>1671.046</td>
      <td>4755355.0</td>
    </tr>
    <tr>
      <th>FB</th>
      <td>167.654</td>
      <td>169.56440</td>
      <td>166.190</td>
      <td>168.510</td>
      <td>19603715.2</td>
    </tr>
    <tr>
      <th>GOOG</th>
      <td>1148.032</td>
      <td>1158.63000</td>
      <td>1134.118</td>
      <td>1150.396</td>
      <td>1321077.0</td>
    </tr>
    <tr>
      <th>TSLA</th>
      <td>288.476</td>
      <td>291.26716</td>
      <td>278.684</td>
      <td>281.904</td>
      <td>15710180.2</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 6: Cho biết giá đóng cửa (close) trung bình, lớn nhất và nhỏ nhất ở mỗi công ty


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>min</th>
      <th>max</th>
    </tr>
    <tr>
      <th>symbol</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAPL</th>
      <td>174.674</td>
      <td>172.50</td>
      <td>175.85</td>
    </tr>
    <tr>
      <th>AMZN</th>
      <td>1671.046</td>
      <td>1625.95</td>
      <td>1696.17</td>
    </tr>
    <tr>
      <th>FB</th>
      <td>168.510</td>
      <td>162.28</td>
      <td>172.51</td>
    </tr>
    <tr>
      <th>GOOG</th>
      <td>1150.396</td>
      <td>1140.99</td>
      <td>1162.03</td>
    </tr>
    <tr>
      <th>TSLA</th>
      <td>281.904</td>
      <td>276.24</td>
      <td>294.79</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 7: Tạo cột parsed_time trong stocks_companies bằng cách đổi thời gian sang định dạng DateTime

# Cho biết kiểu dữ liệu của cột parsed_time

# Hiển thị 5 dòng dữ liệu đầu của stocks_companies


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
      <th>name</th>
      <th>employees</th>
      <th>headquarters_city</th>
      <th>headquarters_state</th>
      <th>parsed_time</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>AMZN</td>
      <td>1655.13</td>
      <td>1674.26</td>
      <td>1651.00</td>
      <td>1671.73</td>
      <td>4974877</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-01-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>AMZN</td>
      <td>1685.00</td>
      <td>1709.43</td>
      <td>1674.36</td>
      <td>1696.17</td>
      <td>6167358</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-04-03</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>AMZN</td>
      <td>1702.95</td>
      <td>1707.80</td>
      <td>1689.01</td>
      <td>1692.43</td>
      <td>3681522</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-05-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>AMZN</td>
      <td>1695.97</td>
      <td>1709.43</td>
      <td>1620.51</td>
      <td>1668.95</td>
      <td>3996001</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-06-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>AMZN</td>
      <td>1667.37</td>
      <td>1669.75</td>
      <td>1620.51</td>
      <td>1625.95</td>
      <td>4957017</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-07-03</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 8: Thêm cột result, nếu giá 'close' >  'open' thì cột result có giá trị 'up', ngược lại 'down'


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>symbol</th>
      <th>open</th>
      <th>high</th>
      <th>low</th>
      <th>close</th>
      <th>volume</th>
      <th>name</th>
      <th>employees</th>
      <th>headquarters_city</th>
      <th>headquarters_state</th>
      <th>parsed_time</th>
      <th>result</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>01-03-19</td>
      <td>AMZN</td>
      <td>1655.13</td>
      <td>1674.26</td>
      <td>1651.00</td>
      <td>1671.73</td>
      <td>4974877</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-01-03</td>
      <td>up</td>
    </tr>
    <tr>
      <th>1</th>
      <td>04-03-19</td>
      <td>AMZN</td>
      <td>1685.00</td>
      <td>1709.43</td>
      <td>1674.36</td>
      <td>1696.17</td>
      <td>6167358</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-04-03</td>
      <td>up</td>
    </tr>
    <tr>
      <th>2</th>
      <td>05-03-19</td>
      <td>AMZN</td>
      <td>1702.95</td>
      <td>1707.80</td>
      <td>1689.01</td>
      <td>1692.43</td>
      <td>3681522</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-05-03</td>
      <td>down</td>
    </tr>
    <tr>
      <th>3</th>
      <td>06-03-19</td>
      <td>AMZN</td>
      <td>1695.97</td>
      <td>1709.43</td>
      <td>1620.51</td>
      <td>1668.95</td>
      <td>3996001</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-06-03</td>
      <td>down</td>
    </tr>
    <tr>
      <th>4</th>
      <td>07-03-19</td>
      <td>AMZN</td>
      <td>1667.37</td>
      <td>1669.75</td>
      <td>1620.51</td>
      <td>1625.95</td>
      <td>4957017</td>
      <td>AMZN</td>
      <td>613300</td>
      <td>Seattle</td>
      <td>WA</td>
      <td>2019-07-03</td>
      <td>down</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>